src/Lower.cpp

```diff
@@ -62,32 +63,37 @@ using std::make_pair;
 
 Stmt lower(const vector<Function> &outputs, const string &pipeline_name, const Target &t, const vector<IRMutator *> &custom_passes) {
 
-    // Compute an environment
+    // Compute an wrapped_environment
     map<string, Function> env;
     for (Function f : outputs) {
         map<string, Function> more_funcs = find_transitive_calls(f);
         env.insert(more_funcs.begin(), more_funcs.end());
     }
 
+    // Create a deep-copy of the Func and substitute in wrapper funcs.
+    const auto &wrapped = wrap_func_calls(outputs, env);
+    const vector<Function> &wrapped_outputs = wrapped.first;
+    const map<string, Function> &wrapped_env = wrapped.second;
+
     // Compute a realization order
-    vector<string> order = realization_order(outputs, env);
+    vector<string> order = realization_order(wrapped_outputs, wrapped_env);
```

Comment:
I think it would read cleaner if you said:
    
    outputs = wrapped_outputs;
    env = wrapped_env;

after this, and then didn't modify the code below. Makes it clear the rest of lowering only looks at the mutated/copied dag.